overlord/servicestate/quota_handlers.go

```diff
@@ -92,46 +93,139 @@ func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
 
 	qc := qcs[0]
 
-	allGrps, err := AllQuotas(st)
+	updated, appsToRestartBySnap, err := quotaStateAlreadyUpdated(t)
 	if err != nil {
 		return err
 	}
 
-	var grp *quota.Group
-	switch qc.Action {
-	case "create":
-		grp, allGrps, err = quotaCreate(st, qc, allGrps)
-	case "remove":
-		grp, allGrps, err = quotaRemove(st, qc, allGrps)
-	case "update":
-		grp, allGrps, err = quotaUpdate(st, qc, allGrps)
-	default:
-		return fmt.Errorf("unknown action %q requested", qc.Action)
+	if !updated {
+		allGrps, err := AllQuotas(st)
+		if err != nil {
+			return err
+		}
+
+		var grp *quota.Group
+		switch qc.Action {
+		case "create":
+			grp, allGrps, err = quotaCreate(st, qc, allGrps)
+		case "remove":
+			grp, allGrps, err = quotaRemove(st, qc, allGrps)
+		case "update":
+			grp, allGrps, err = quotaUpdate(st, qc, allGrps)
+		default:
+			return fmt.Errorf("unknown action %q requested", qc.Action)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		// ensure service and slices on disk and their states are updated
+		opts := &ensureSnapServicesForGroupOptions{
+			allGrps: allGrps,
+		}
+		appsToRestartBySnap, err = ensureSnapServicesForGroup(st, t, grp, opts)
+		if err != nil {
+			return err
+		}
+
+		// after we have made all the persistent modifications
+		// to disk and state, remember that so to avoid
+		// failing redoing the non-idempotent parts of the
+		// task if snapd gets restarted before the end of the task.
+		// What remains for this task handler is just to
+		// restart services which will happen regardless if we
+		// get rebooted after unlocking the state - if we got
+		// rebooted before unlocking the state, none of the
+		// changes we made to state would be persisted and we
+		// would run through everything above here again, but
+		// the second time around EnsureSnapServices would end
+		// up doing nothing since it is idempotent.  So in the
+		// rare case that snapd gets restarted but is not a
+		// reboot also remember which service do need
+		// restarting. There is a small chance that services
+		// will be restarted again but is preferable to the
+		// quota not applying to them.
+		if err := rememberQuotaStateUpdated(t, appsToRestartBySnap); err != nil {
```

Comment:
Maybe the comment should mention that because this is internally using task progress adapter locked, there is no possibility of the group changes in the state will be committed to the disk before rememberQuotaStateUpdated runs.